Update new package README template

mjwolf opened the pull request:
Update the README templates and generator used in new package creation to match the new standardized README structure. The updated template should be better at helping users write the readme for new packages, by explaining what is needed in each section of the template.

This also makes a few other small changes which will make it easier for new users to create and use integration packages. It moves the generated readme to `_dev/build/docs`, automatically adds a "Do not modify" message to the README generated during `build`, and adds a default sample-event (which is needed to build the update readme template, since it now includes `{{event}}`).

mjwolf commented:
The new README template in Kibana, with no changes made to it:

![mw-dev-e3609a kb canadacentral azure staging elastic cloud_app_integrations_detail_hello-0 0 9_overview](https://github.com/user-attachments/assets/6b41744d-2eb9-4f9c-9df4-d02fbec48296)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-package/builds/5593)
* Commit: 9c517321a1c5cd973143dc05841653cf67b9983a

### Failed CI Steps
* [:linux: Run unit tests](https://buildkite.com/elastic/elastic-package/builds/5593#0197d240-b83c-45af-a9f8-c14e125d9e24)

<!--buildkite-pr-comment-elastic-package
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5593","number":5593,"commit":"9c517321a1c5cd973143dc05841653cf67b9983a"}],"number":5593}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-package/builds/5600)
* Commit: c2323d263fd07d6d6bfebfdec2c5e47c3afb8218

### Failed CI Steps
* [:go: Run check-static](https://buildkite.com/elastic/elastic-package/builds/5600#0197d6b6-44cd-4143-8929-a49ff29c103f)

### History
* :broken_heart: [Build #5593](https://buildkite.com/elastic/elastic-package/builds/5593) failed 9c517321a1c5cd973143dc05841653cf67b9983a

<!--buildkite-pr-comment-elastic-package
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5600","number":5600,"commit":"c2323d263fd07d6d6bfebfdec2c5e47c3afb8218"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5593","number":5593,"commit":"9c517321a1c5cd973143dc05841653cf67b9983a"}],"number":5600}
buildkite-pr-comment-->

mjwolf commented on internal/docs/readme.go:
```diff
@@ -204,6 +209,9 @@ func renderReadme(fileName, packageRoot, templatePath string, linksMap linkMap)
 			}
 			return linksMap.RenderLink(args[0], options)
 		},
+		"header": func(args ...string) (string, error) {
```
This depends on the user leaving ``{{header}}` in the readme in order to add the "Do not modify" string to the generated files.

I could also inject the string into all generated READMEs, regardless of if this {{header}} is in it, but I don't know if this would cause problems with other package types, so I didn't do that. If it won't cause problems, I could change to add to all generated files.

jsoriano replied:
Yep, it could make sense to inject it in all generated files, but we would need to check that validation doesn't fail for packages that don't include this header.

mjwolf commented:
> Could you modify one of the packages under `test/packages` to see the expected usage and result?

The `panw` integration has already been published with the README updated to use the new template: https://github.com/elastic/integrations/tree/main/packages/panw/docs

I can also modify one of the test packages in elastic-package, so there's an example of new template here.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-package/builds/5617)
* Commit: ecf9bed6254b500daa1823cb1bc1cbf312fbf134

### Failed CI Steps
* [:go: Integration test: system](https://buildkite.com/elastic/elastic-package/builds/5617#0197eb00-61b3-4b7e-8e04-dbe01f4a30b5)

### History
* :broken_heart: [Build #5600](https://buildkite.com/elastic/elastic-package/builds/5600) failed c2323d263fd07d6d6bfebfdec2c5e47c3afb8218
* :broken_heart: [Build #5593](https://buildkite.com/elastic/elastic-package/builds/5593) failed 9c517321a1c5cd973143dc05841653cf67b9983a

<!--buildkite-pr-comment-elastic-package
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5617","number":5617,"commit":"ecf9bed6254b500daa1823cb1bc1cbf312fbf134"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5600","number":5600,"commit":"c2323d263fd07d6d6bfebfdec2c5e47c3afb8218"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5593","number":5593,"commit":"9c517321a1c5cd973143dc05841653cf67b9983a"}],"number":5617}
buildkite-pr-comment-->

mjwolf commented:
I've changed the readme in the `test/packages/parallel/apache` package to use the new format.  This isn't used in the real, published `apache` integration yet though. I've just added it here so there's an example of the new format in elastic-package.

jsoriano commented on test/packages/parallel/apache/_dev/build/docs/README.md:
```diff
@@ -1,34 +1,129 @@
+Access and error logs data stream read from a logfile. You will configure Apache server to write log files to a location that is readable by elastic-agent, and 
+elastic-agent will monitor and ingest events written to these files.
 
-{{fields "access"}}
```
It looks like `fields` are not rendered anymore. In the rendered README I can only see `{{fields}}`, in https://github.com/elastic/elastic-package/pull/2707/files#diff-21c266b51e67f0cc833e6b48a09867796a652d1b2d9abe69119b3f36d7975d99R114.

kgeller commented on internal/packages/archetype/package.go:
```diff
@@ -54,6 +54,14 @@ func createPackageInDir(packageDescriptor PackageDescriptor, cwd string) error {
 		return fmt.Errorf("can't render package README: %w", err)
 	}
 
+	if packageDescriptor.Manifest.Type == "integration" {
```
yeah I think so, dropping the if check

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-package/builds/5853)
* Commit: d7fd5dfa10f0126b267c710105ead2c5ce8cfaa0

### Failed CI Steps
* [:go: Run check-static](https://buildkite.com/elastic/elastic-package/builds/5853#0198621c-5aef-4420-b66c-f8c8fdc8d9d6)

### History
* :broken_heart: [Build #5844](https://buildkite.com/elastic/elastic-package/builds/5844) failed d901942a5d977eb86ad2a859ddc8fea87fa0da43
* :broken_heart: [Build #5617](https://buildkite.com/elastic/elastic-package/builds/5617) failed ecf9bed6254b500daa1823cb1bc1cbf312fbf134
* :broken_heart: [Build #5600](https://buildkite.com/elastic/elastic-package/builds/5600) failed c2323d263fd07d6d6bfebfdec2c5e47c3afb8218
* :broken_heart: [Build #5593](https://buildkite.com/elastic/elastic-package/builds/5593) failed 9c517321a1c5cd973143dc05841653cf67b9983a

<!--buildkite-pr-comment-elastic-package
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5853","number":5853,"commit":"d7fd5dfa10f0126b267c710105ead2c5ce8cfaa0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5844","number":5844,"commit":"d901942a5d977eb86ad2a859ddc8fea87fa0da43"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5617","number":5617,"commit":"ecf9bed6254b500daa1823cb1bc1cbf312fbf134"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5600","number":5600,"commit":"c2323d263fd07d6d6bfebfdec2c5e47c3afb8218"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5593","number":5593,"commit":"9c517321a1c5cd973143dc05841653cf67b9983a"}],"number":5853}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-package/builds/5854)
* Commit: 031eeafa04471b61382995b50797fd008a923fad

### Failed CI Steps
* [:linux: Run unit tests](https://buildkite.com/elastic/elastic-package/builds/5854#01986220-910d-4afd-9195-1de5c1b7c5e3)

### History
* :broken_heart: [Build #5853](https://buildkite.com/elastic/elastic-package/builds/5853) failed d7fd5dfa10f0126b267c710105ead2c5ce8cfaa0
* :broken_heart: [Build #5844](https://buildkite.com/elastic/elastic-package/builds/5844) failed d901942a5d977eb86ad2a859ddc8fea87fa0da43
* :broken_heart: [Build #5617](https://buildkite.com/elastic/elastic-package/builds/5617) failed ecf9bed6254b500daa1823cb1bc1cbf312fbf134
* :broken_heart: [Build #5600](https://buildkite.com/elastic/elastic-package/builds/5600) failed c2323d263fd07d6d6bfebfdec2c5e47c3afb8218
* :broken_heart: [Build #5593](https://buildkite.com/elastic/elastic-package/builds/5593) failed 9c517321a1c5cd973143dc05841653cf67b9983a

<!--buildkite-pr-comment-elastic-package
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5854","number":5854,"commit":"031eeafa04471b61382995b50797fd008a923fad"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5853","number":5853,"commit":"d7fd5dfa10f0126b267c710105ead2c5ce8cfaa0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5844","number":5844,"commit":"d901942a5d977eb86ad2a859ddc8fea87fa0da43"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5617","number":5617,"commit":"ecf9bed6254b500daa1823cb1bc1cbf312fbf134"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5600","number":5600,"commit":"c2323d263fd07d6d6bfebfdec2c5e47c3afb8218"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5593","number":5593,"commit":"9c517321a1c5cd973143dc05841653cf67b9983a"}],"number":5854}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/elastic-package/builds/5858)
* Commit: 80d6c245b1a80d5032ec41cbca3745986eb9161b

### History
* :broken_heart: [Build #5857](https://buildkite.com/elastic/elastic-package/builds/5857) failed 402120c15b76eadd6b02629065ae499530579f16
* :broken_heart: [Build #5856](https://buildkite.com/elastic/elastic-package/builds/5856) failed e03d736ad5f789482d191a8ebf42889c0b76c6c0
* :broken_heart: [Build #5855](https://buildkite.com/elastic/elastic-package/builds/5855) failed 962358e96abd5d063724497a6e89588d01896d9d
* :broken_heart: [Build #5854](https://buildkite.com/elastic/elastic-package/builds/5854) failed 031eeafa04471b61382995b50797fd008a923fad
* :broken_heart: [Build #5853](https://buildkite.com/elastic/elastic-package/builds/5853) failed d7fd5dfa10f0126b267c710105ead2c5ce8cfaa0

<!--buildkite-pr-comment-elastic-package
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5858","number":5858,"commit":"80d6c245b1a80d5032ec41cbca3745986eb9161b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5857","number":5857,"commit":"402120c15b76eadd6b02629065ae499530579f16"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5856","number":5856,"commit":"e03d736ad5f789482d191a8ebf42889c0b76c6c0"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5855","number":5855,"commit":"962358e96abd5d063724497a6e89588d01896d9d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5854","number":5854,"commit":"031eeafa04471b61382995b50797fd008a923fad"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5853","number":5853,"commit":"d7fd5dfa10f0126b267c710105ead2c5ce8cfaa0"}],"number":5858}
buildkite-pr-comment-->

jsoriano commented on internal/packages/archetype/_static/package-docs-readme.md.tmpl:
```diff
@@ -1,84 +1,105 @@
 
-<!-- The {{.Manifest.Title}} integration allows you to monitor {name of service}. {name of service} is {describe service}.
+<!-- Do not remove header -->
+{{ `{{header}}` }}
```
Adding test that would have detected this issue https://github.com/elastic/elastic-package/pull/2796.

kgeller replied:
Thanks for adding that test! 

I intended to drop that function all together, but I somehow missed it being called in the template 🥲 

Pushing up some changes now, but I also tested it locally as well with the create and then the build, and it is now successful

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/elastic-package/builds/5873)
* Commit: 4688fad502cdcf1025962d0b4592a97c72213225

### History
* :green_heart: [Build #5858](https://buildkite.com/elastic/elastic-package/builds/5858) succeeded 80d6c245b1a80d5032ec41cbca3745986eb9161b
* :broken_heart: [Build #5857](https://buildkite.com/elastic/elastic-package/builds/5857) failed 402120c15b76eadd6b02629065ae499530579f16
* :broken_heart: [Build #5856](https://buildkite.com/elastic/elastic-package/builds/5856) failed e03d736ad5f789482d191a8ebf42889c0b76c6c0
* :broken_heart: [Build #5855](https://buildkite.com/elastic/elastic-package/builds/5855) failed 962358e96abd5d063724497a6e89588d01896d9d
* :broken_heart: [Build #5854](https://buildkite.com/elastic/elastic-package/builds/5854) failed 031eeafa04471b61382995b50797fd008a923fad
* :broken_heart: [Build #5853](https://buildkite.com/elastic/elastic-package/builds/5853) failed d7fd5dfa10f0126b267c710105ead2c5ce8cfaa0

<!--buildkite-pr-comment-elastic-package
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5873","number":5873,"commit":"4688fad502cdcf1025962d0b4592a97c72213225"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5858","number":5858,"commit":"80d6c245b1a80d5032ec41cbca3745986eb9161b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5857","number":5857,"commit":"402120c15b76eadd6b02629065ae499530579f16"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5856","number":5856,"commit":"e03d736ad5f789482d191a8ebf42889c0b76c6c0"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5855","number":5855,"commit":"962358e96abd5d063724497a6e89588d01896d9d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5854","number":5854,"commit":"031eeafa04471b61382995b50797fd008a923fad"}],"number":5873}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/elastic-package/builds/5887)
* Commit: 0e6cc140ed607e36a0a597b1ff51b7179bddc989

### History
* :broken_heart: [Build #5885](https://buildkite.com/elastic/elastic-package/builds/5885) failed 3977bc8e904be16e5402475ae0627312dd999980
* :green_heart: [Build #5873](https://buildkite.com/elastic/elastic-package/builds/5873) succeeded 4688fad502cdcf1025962d0b4592a97c72213225
* :green_heart: [Build #5858](https://buildkite.com/elastic/elastic-package/builds/5858) succeeded 80d6c245b1a80d5032ec41cbca3745986eb9161b
* :broken_heart: [Build #5857](https://buildkite.com/elastic/elastic-package/builds/5857) failed 402120c15b76eadd6b02629065ae499530579f16
* :broken_heart: [Build #5856](https://buildkite.com/elastic/elastic-package/builds/5856) failed e03d736ad5f789482d191a8ebf42889c0b76c6c0

<!--buildkite-pr-comment-elastic-package
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5887","number":5887,"commit":"0e6cc140ed607e36a0a597b1ff51b7179bddc989"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5885","number":5885,"commit":"3977bc8e904be16e5402475ae0627312dd999980"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5873","number":5873,"commit":"4688fad502cdcf1025962d0b4592a97c72213225"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5858","number":5858,"commit":"80d6c245b1a80d5032ec41cbca3745986eb9161b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5857","number":5857,"commit":"402120c15b76eadd6b02629065ae499530579f16"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-package/builds/5856","number":5856,"commit":"e03d736ad5f789482d191a8ebf42889c0b76c6c0"}],"number":5887}
buildkite-pr-comment-->

jsoriano commented on internal/packages/archetype/_static/package-docs-readme.md.tmpl:
```diff
@@ -1,84 +1,118 @@
+#### {data stream name} fields
+
+To include a generated list of fields from the `fields/` directory, uncomment and use:
+(Remove the spaces between curly braces when using)
+{ { fields "data_stream_name" } }
```
We can directly generate the expected curly braces.
```suggestion
{{ "{{ fields \"data_stream_name\" }}" }}
```

jsoriano replied:
> Are you saying we should use the Go template comments for both the scenarios above, and then try and modify the readme generation (`elastic-package build`) to strip out all comments? I tried this, but the go template comments get stripped out for both readme's when I run `create package` (unless there's something with the rendering I can also update).

Yes, but we could use different delimiters for each step. So let's say that we use `<<`/`>>` as delimiters for the `create` step, and `{{`/`}}` for the build step, then we could have comments with `{{/*`/`*/}}` that would stay in the template, but won't be visible in the final README.

I see how this complicates things and may require even a small refactor to allow the different delimiters, we may leave this for a different PR. I think it is worth it because this way comments can stay in the template and are easier to incorporate by package developers.

jsoriano commented on internal/packages/archetype/_static/package-docs-readme.md.tmpl:
```diff
@@ -1,84 +1,118 @@
+<!--
+
+To include a sample event from `sample_event.json`, uncomment and use:
+(Remove the spaces between curly braces when using)
+{ { event "data_stream_name" } }
```
```suggestion
{{ "{{ event \"data_stream_name\" }}" }}
```

jsoriano reviewed (approved):
LGTM, the new content is much more useful and complete, thanks!

We can leave the refactor of comments/delimiters in templates for a different iteration.
